db/blip_handler.go

```diff
@@ -792,65 +823,107 @@ func (bh *blipHandler) handleGetAttachment(rq *blip.Message) error {
 func (bh *blipHandler) downloadOrVerifyAttachments(sender *blip.Sender, body Body, minRevpos int, docID string) error {
 	return bh.db.ForEachStubAttachment(body, minRevpos,
 		func(name string, digest string, knownData []byte, meta map[string]interface{}) ([]byte, error) {
-			if knownData != nil {
-				// If I have the attachment already I don't need the client to send it, but for
-				// security purposes I do need the client to _prove_ it has the data, otherwise if
-				// it knew the digest it could acquire the data by uploading a document with the
-				// claimed attachment, then downloading it.
-				base.DebugfCtx(bh.loggingCtx, base.KeySync, "    Verifying attachment %q for doc %s (digest %s)", base.UD(name), base.UD(docID), digest)
-				nonce, proof := GenerateProofOfAttachment(knownData)
-				outrq := blip.NewRequest()
-				outrq.Properties = map[string]string{BlipProfile: MessageProveAttachment, ProveAttachmentDigest: digest}
-				outrq.SetBody(nonce)
-				if !bh.sendBLIPMessage(sender, outrq) {
-					return nil, ErrClosedBLIPSender
-				}
-				if body, err := outrq.Response().Body(); err != nil {
-					base.WarnfCtx(bh.loggingCtx, "Error returned for proveAttachment message for doc %s (digest %s).  Error: %v", base.UD(docID), digest, err)
-					return nil, err
-				} else if string(body) != proof {
-					base.WarnfCtx(bh.loggingCtx, "Incorrect proof for attachment %s : I sent nonce %x, expected proof %q, got %q", digest, base.MD(nonce), base.MD(proof), base.MD(string(body)))
-					return nil, base.HTTPErrorf(http.StatusForbidden, "Incorrect proof for attachment %s", digest)
-				} else {
-					base.InfofCtx(bh.loggingCtx, base.KeySync, "proveAttachment successful for doc %s (digest %s)", base.UD(docID), digest)
-				}
-				return nil, nil
-			} else {
-				// If I don't have the attachment, I will request it from the client:
-				base.DebugfCtx(bh.loggingCtx, base.KeySync, "    Asking for attachment %q for doc %s (digest %s)", base.UD(name), base.UD(docID), digest)
-				outrq := blip.NewRequest()
-				outrq.Properties = map[string]string{BlipProfile: MessageGetAttachment, GetAttachmentDigest: digest}
-				if isCompressible(name, meta) {
-					outrq.Properties[BlipCompress] = "true"
-				}
-				if !bh.sendBLIPMessage(sender, outrq) {
-					return nil, ErrClosedBLIPSender
-				}
-				attBody, err := outrq.Response().Body()
-				if err != nil {
-					return nil, err
-				}
 
-				lNum, metaLengthOK := meta["length"].(json.Number)
-				metaLength, err := lNum.Int64()
-				if err != nil {
-					return nil, err
-				}
+			haveAttachment := knownData != nil
 
-				// Verify that the attachment we received matches the metadata stored in the document
-				if !metaLengthOK || len(attBody) != int(metaLength) || Sha1DigestKey(attBody) != digest {
-					return nil, base.HTTPErrorf(http.StatusBadRequest, "Incorrect data sent for attachment with digest: %s", digest)
-				}
+			if !haveAttachment {
```

Comment:
Since we're not using haveAttachment elsewhere any more, could simplify this.
```suggestion
			if knownData == nil {
```